subscriptions/subscriptions-impl/src/main/java/com/duckduckgo/subscriptions/impl/SubscriptionsManager.kt

```diff
@@ -93,25 +108,82 @@ class RealSubscriptionsManager @Inject constructor(
     private val authDataStore: AuthDataStore,
     private val billingClientWrapper: BillingClientWrapper,
     private val context: Context,
+    @AppCoroutineScope private val coroutineScope: CoroutineScope,
+    private val dispatcherProvider: DispatcherProvider,
 ) : SubscriptionsManager {
 
     private val adapter = Moshi.Builder().build().adapter(ResponseError::class.java)
 
+    private val _currentPurchaseState = MutableSharedFlow<CurrentPurchase>()
+    override val currentPurchaseState = _currentPurchaseState.asSharedFlow()
+
     private val _isSignedIn = MutableStateFlow(isUserAuthenticated())
     override val isSignedIn = _isSignedIn.asStateFlow()
 
+    private val _hasSubscription = MutableStateFlow(false)
+    override val hasSubscription = _hasSubscription.asStateFlow()
+
     private fun isUserAuthenticated(): Boolean = !authDataStore.accessToken.isNullOrBlank() && !authDataStore.authToken.isNullOrBlank()
 
+    init {
+        coroutineScope.launch(dispatcherProvider.io()) {
+            _hasSubscription.emit(hasSubscription())
+        }
+
+        coroutineScope.launch(dispatcherProvider.io()) {
+            billingClientWrapper.purchaseState.collect {
+                when (it) {
+                    is PurchaseState.Purchased -> {
+                        checkPurchase()
+                    }
+                    is PurchaseState.Canceled -> {
+                        checkPurchase()
+                    }
+                    else -> {
+                        // NOOP
+                    }
+                }
+            }
+        }
+    }
+
     override suspend fun signOut() {
         authDataStore.authToken = ""
         authDataStore.accessToken = ""
         _isSignedIn.emit(isUserAuthenticated())
+        _hasSubscription.emit(false)
     }
 
-    override suspend fun hasSubscription(): Boolean {
+    private suspend fun checkPurchase() {
+        _currentPurchaseState.emit(CurrentPurchase.InProgress)
+        delay(500)
+        var retries = 1
+        var data = hasSubscription()
+        while (!data && retries <= 3) {
+            delay(500L * retries)
+            data = hasSubscription()
+            retries++
+        }
+        if (data) {
+            _currentPurchaseState.emit(CurrentPurchase.Success)
+            _hasSubscription.emit(true)
+        } else {
+            _hasSubscription.emit(false)
+            _currentPurchaseState.emit(CurrentPurchase.Failure("An error happened, try again"))
+        }
+    }
+
+    private suspend fun hasSubscription(): Boolean {
         return when (val result = getSubscriptionData()) {
-            is Success -> return result.entitlements.isNotEmpty()
-            is Failure -> false
+            is Success -> {
+                val isSubscribed = result.entitlements.isNotEmpty()
+                _hasSubscription.emit(isSubscribed)
```

Comment:
I think we should remove `_hasSubscription.emit(isSubscribed)` from this function. All usages of this function already emits the value received here. It is a bit confusing that this also emits the same value.

Reply:
Done!